lib/dyph3/two_way_differs/heckel_diff.rb

```diff
@@ -1,85 +1,178 @@
+            changes_ranges << [:add, left_lo + 1, left_lo, right_lo + 1, right_hi + 1]
+          end
+          changes_ranges
         end
 
-        def self.move_to_prev_difference(text_a, text_b, a0, b0, a1, b1)
-          # loop from a1 and b1's new positions down towards a0, b0.  stop when there is a change.  This gives the bounds of the change as [a0,a1] and [b0, b1]
-          while a0 <= a1 && b0 <= b1
-            break if text_a[a1] != text_b[b1]   # a change is found on lines a1 and b1.  break out of loop.
-            a1 -= 1
-            b1 -= 1
+        def self.convert_to_typed_ouput(heckel_diff, old_text_array, new_text_array)
+          chunks = heckel_diff.map { |block| TwoWayChunk.new(block) }
+
+          old_index = 0
+          old_text = []
+          new_index = 0
+          new_text = []
+          chunks.each do |chunk|
+            old_iteration = 0
+            while old_index + old_iteration < chunk.left_lo - 1     # chunk indexes are from 1
+              old_text << TextNode.new(text: old_text_array[old_index + old_iteration], row: new_index + old_iteration)
+              old_iteration += 1
+            end
+
+            new_iteration = 0
+            while new_index + new_iteration < chunk.right_lo - 1     # chunk indexes are from 1
+              new_text << TextNode.new(text: new_text_array[new_index + new_iteration], row: old_index + new_iteration)
+              new_iteration += 1
+            end
+
+            old_index += old_iteration
+            new_index += new_iteration
+
+            while old_index <= chunk.left_hi - 1     # chunk indexes are from 1
+              old_text << old_text_array[old_index]
+              old_index += 1
+            end
+
+            while new_index <= chunk.right_hi - 1     # chunk indexes are from 1
+              new_text << new_text_array[new_index]
+              new_index += 1
+            end
+          end
+
+          iteration = 0
+          while old_index + iteration < old_text_array.length
+            old_text << TextNode.new(text: old_text_array[old_index + iteration], row: new_index + iteration)
+            iteration += 1
           end
-          [a1, b1]
+
+          iteration = 0
+          while new_index + iteration < new_text_array.length
+            new_text << TextNode.new(text: new_text_array[new_index + iteration], row: old_index + iteration)
+            iteration += 1
+          end
+
+          { old_text: old_text, new_text: new_text}
         end
     end
+
+    class TextNode
+      attr_accessor :text, :row
+
+      def initialize(text:, row:)
+        @text = text
+        @row  = row
+      end
+    end
+
+    class TwoWayChunk
+      attr_reader :action, :left_lo, :left_hi, :right_lo, :right_hi
+      def initialize(raw_chunk)
+        @action   = raw_chunk[0]
+        @left_lo  = raw_chunk[1]
+        @left_hi  = raw_chunk[2]
+        @right_lo = raw_chunk[3]
+        @right_hi = raw_chunk[4]
+      end
+    end
+
```

Comment:
`TextNode` and `TwoWayChunk` could easily be good ol' structs instead, with their sibling `ChangeData`.